crates/oxc_linter/src/rules/jsx_a11y/role_has_required_aria_props.rs

```diff
@@ -0,0 +1,142 @@
+use crate::{context::LintContext, rule::Rule, utils::has_jsx_prop_lowercase, AstNode};
+use oxc_ast::{
+    ast::{JSXAttributeItem, JSXAttributeValue},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::{self, Error},
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+use phf::phf_map;
+
+#[derive(Debug, Error, Diagnostic)]
+#[error(
+    "eslint-plugin-jsx-a11y(role-has-required-aria-props): `{role}` role is missing required aria props `{props}`."
+)]
+#[diagnostic(
+    severity(warning),
+    help("Add missing aria props `{props}` to the element with `{role}` role.")
+)]
+struct RoleHasRequiredAriaPropsDiagnostic {
+    #[label]
+    pub span: Span,
+    pub role: String,
+    pub props: String,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct RoleHasRequiredAriaProps;
+declare_oxc_lint!(
+    /// ### What it does
+    /// Enforces that elements with ARIA roles must have all required attributes for that role.
+    ///
+    /// ### Why is this bad?
+    /// Certain ARIA roles require specific attributes to express necessary semantics for assistive technology.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// <div role="checkbox" />
+    ///
+    /// // Good
+    /// <div role="checkbox" aria-checked="false" />
+    /// ````
+    RoleHasRequiredAriaProps,
+    correctness
+);
+
+static ROLE_TO_REQUIRED_ARIA_PROPS: phf::Map<&'static str, &'static str> = phf_map! {
+    "checkbox" => "aria-checked",
+    "radio" => "aria-checked",
+    "combobox" => "aria-controls, aria-expanded",
+    "tab" => "aria-selected",
+    "slider" => "aria-valuemax, aria-valuemin, aria-valuenow",
+    "scrollbar" => "aria-valuemax, aria-valuemin, aria-valuenow, aria-orientation, aria-controls",
+    "heading" => "aria-level",
+    "option" => "aria-selected",
+};
+
+impl Rule for RoleHasRequiredAriaProps {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::JSXOpeningElement(jsx_el) = node.kind() {
+            let Some(role_prop) = has_jsx_prop_lowercase(jsx_el, "role") else { return };
+            let JSXAttributeItem::Attribute(attr) = role_prop else { return };
+            let Some(JSXAttributeValue::StringLiteral(role_values)) = &attr.value else { return };
+            let roles = role_values.value.split_whitespace();
+            for role in roles {
+                let required_props = match ROLE_TO_REQUIRED_ARIA_PROPS.get(role) {
+                    Some(required_props) => required_props,
+                    None => continue,
+                };
+                for prop in required_props.split(',').map(str::trim) {
```

Comment:
shall we move this to compile time rather than runtime? it should help improve performance.

if we change the `ROLE_TO_REQUIRED_ARIA_PROPS` to be `phf::Map<&'static str, phf::Set<&'static str>>` we should be good to go

Reply:
Thanks for the multiple reviews!
At https://github.com/oxc-project/oxc/pull/1881/commits/7a2f75c47dcddc57bd421ad6819d87238dfed146 ←, at compile time, the required properties for each role are saved as a string The following is a list of the properties that are required for each role at compile time!

Reply:
🙏 thanks!